components/datetime/src/provider/neo.rs

```diff
@@ -0,0 +1,128 @@
+///
+/// This uses an auxiliary key for length.
+#[icu_provider::data_struct(marker(YearSymbolsV1Marker, "datetime/symbols/years@1"))]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+pub enum YearSymbolsV1<'data> {
+    /// This calendar uses eras with numeric years, this stores the era names mapped from
+    /// era code to the name
+    Eras(#[cfg_attr(feature = "serde", serde(borrow))] ZeroMap<'data, UnvalidatedStr, str>),
+    /// This calendar is cyclic (Chinese, Dangi), so it uses cyclic year names without any eras
+    Cyclic(#[cfg_attr(feature = "serde", serde(borrow))] VarZeroVec<'data, str>),
+}
+
+/// Symbols used for representing the month name
+///
+/// This uses an auxiliary key for length.
+#[icu_provider::data_struct(marker(MonthSymbolsV1Marker, "datetime/symbols/months@1"))]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+pub enum MonthSymbolsV1<'data> {
+    /// Month codes M01, M02, M03, .. (can allow for M13 onwards)
+    ///
+    /// Found for solar and pure lunar calendars
+    Numeric(#[cfg_attr(feature = "serde", serde(borrow))] VarZeroVec<'data, str>),
+    /// Month code map that can handle arbitrary month codes including leap months
+    ///
+    /// Found for lunisolar and lunisidereal calendars
+    Map(#[cfg_attr(feature = "serde", serde(borrow))] ZeroMap<'data, UnvalidatedStr, str>),
+}
+
+/// Symbols that can be stored as a simple linear array.
+///
+/// - For weekdays, element 0 is Monday
+/// - For dayperiods, the elements are in order: AM, PM, (noon), (midnight), where the latter two are optional
+/// - For day names element 0 is the first day of the month
+///
+/// This uses an auxiliary key for length.
+#[icu_provider::data_struct(
+    marker(WeekdaySymbolsV1Marker, "datetime/symbols/weekdays@1"),
+    marker(DayPeriodSymbolsV1Marker, "datetime/symbols/dayperiods@1"),
+    marker(DaySymbolsV1Marker, "datetime/symbols/days@1")
+)]
+#[derive(Debug, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_datetime::provider::neo),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[yoke(prove_covariance_manually)]
+pub struct LinearSymbolsV1<'data> {
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    symbols: VarZeroVec<'data, str>,
+}
+
+/// The default per-length patterns associated with dates and times
+///
+/// This uses an auxiliary key for length. time@1 additionally uses
+/// the auxiliary key for representing hour cycle preferences.
```

Comment:
I just realized that a risk is that this gets us a bit close to the precipice of implementing a horizontal fallback that Mihai has advised against. For example, if en-GB has `format` and en has both `format` and `standalone`, and a user in en-GB asks for standalone, we are supposed to load en-GB `format` and not en `standalone`.

A few choices:

1. Always populate the data and rely on deduplication to fix it up
2. Make some of these have an `Option` field
3. Make the data provider aware of keys that are present but empty (blocking fallback)
4. Add metadata to the key describing to the data provider how to perform fallback and leave it up to the data provider implementation to do it correctly

Reply:
I was under the impression we had discussed this already: this was one of my major concerns about auxiliary keys. I thought we were planning on exposing the ability to look up which aux keys are available or something like that, in a way that respects fallbacking.

If we don't have that we're going to have more problems with aux keys in currencies.

Reply:
Listing aux keys was the plan iirc, however I don't recall that we planned on that interacting with fallback, i.e. including aux keys for parents. This wouldn't be hard to do, but would require allocating a set, whereas listing aux keys without fallback can be an iterator.